rmw_cyclonedds_cpp/src/serdata.cpp

```diff
@@ -685,11 +749,351 @@ static std::string get_type_name(const char * type_support_identifier, void * ty
+  }
+
+  dds_free_typeinfo(type_info);
+  dds_delete_topic_descriptor(desc);
+  dds_dynamic_type_unref(&dt);
+}
+
+template<typename MembersType>
+static bool construct_dds_dynamic_type(
+  dds_dynamic_type_t * dstruct, dds_entity_t dds_ppant, const MembersType * members)
+{
+  assert(members);
+  assert(dds_ppant);
+ 
+  dds_return_t ret;
+
+  for (uint32_t i = 0; i < members->member_count_; ++i)
+  {
+    const auto * member = members->members_ + i; 
+    switch (member->type_id_) 
+    {
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BOOL:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_BOOLEAN);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BYTE:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_BYTE);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT8:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_UINT8);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_CHAR:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_CHAR8);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT8:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_INT8);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT32:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_FLOAT32);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT64:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_FLOAT64);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT16:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_INT16);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT16:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_UINT16);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT32:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_INT32);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT32:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_UINT32);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT64:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_INT64);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT64:
+        dynamic_type_add_member(dstruct, dds_ppant, member, DDS_DYNAMIC_UINT64);
+        break;
+      case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
+        {
+          dds_dynamic_type_t ddt;
+          if (member->is_upper_bound_)
+          {
+            ddt = dds_dynamic_type_create(dds_ppant,
+              get_dynamic_type_descriptor(DDS_DYNAMIC_STRING8, nullptr, 1, (uint32_t*)(&member->array_size_), {})); 
```

Comment:
Here the same thing I mentioned above about `size_t` vs `uint32_t` applies as well.